### PR TITLE
Quick fixes for snooze/dismiss functionality in tasks API

### PR DESCRIPTION
**Pull request by joelclimbsthings**
Quick fixes to snooze and dismiss endpoints on task list API. 

### Detailed test instructions:

-   Checkout branch
-   Ensure one of the tasks `isDismissible` by adding `'is_dismissable' => true` to a task file such as [this one](https://github.com/woocommerce/woocommerce-admin/blob/31e0e4988171cf997b2ab995d0475d4b780b8a20/src/Features/OnboardingTasks/Tasks/Appearance.php?plain=1#L26). 
- Go to the homescreen and dismiss the task.
- Ensure it disappears is added to the option `woocommerce_task_list_dismissed_tasks`

### CODE REVIEW

File: src/Features/OnboardingTasks/TaskLists.php [hunk 1]
```diff
@@ -165,7 +165,8 @@ public static function get_task( $id, $task_list_id = null ) {
 		$tasks_to_search = $task_list ? $task_list['tasks'] : array_reduce(
 			self::get_lists(),
 			function ( $all, $curr ) {
-				return array_merge( $all, $curr['tasks'] );
+				$json = $curr->get_json();
+				return array_merge( $all, $json['tasks'] );
```
Comment (louwie17) [inline src/Features/OnboardingTasks/TaskLists.php:169]: @joelclimbsthings instead of converting it to `json` and then getting the tasks, should we instead expose either the `tasks` variable (making it public) or creating a `get_tasks` method within `TaskList.php`?

Comment (joshuatf) [reply src/Features/OnboardingTasks/TaskLists.php]: Agreed.  Keep it in model form as long as possible before returning as JSON.

**Review by joshuatf — COMMENTED** [review]
Ah, I didn't realize you were working on this one, @joelclimbsthings.  I also have a PR that addresses these problems and fixes up some duplication in the endpoints - https://github.com/woocommerce/woocommerce-admin/pull/7728

File: src/Features/OnboardingTasks/TaskLists.php (hunk 1 above)
Comment (joelclimbsthings) [reply src/Features/OnboardingTasks/TaskLists.php]: Totally agreed @joshuatf @louwie17 . This led to doing more work on the endpoints than intended in this PR, but done in b37f537cbadb3dc2ff49ba08606b7f059281a004.

**Comment by joelclimbsthings** [comment]
>  Ah, I didn't realize you were working on this one, @joelclimbsthings. I also have a PR that addresses these problems and fixes up some duplication in the endpoints

Yeah, I was afraid that was the case. I find myself frequently having the thought "I wonder if Josh is already working on this?" while addressing these things, which is why I decided to get a PR up quickly. We'll have to find a way to coordinate a bit better.

**Comment by joelclimbsthings** [comment]
@joshuatf @louwie17 I've made the suggested changes, let me know what you think!

**Review by joshuatf — COMMENTED** [review]
Thanks for the changes, @joelclimbsthings.  Undoing dismiss and snooze is still broken in this branch.  (See https://github.com/woocommerce/woocommerce-admin/pull/7728 for fixes there)

>Yeah, I was afraid that was the case. I find myself frequently having the thought "I wonder if Josh is already working on this?" while addressing these things, which is why I decided to get a PR up quickly. We'll have to find a way to coordinate a bit better.

I think the answer to this is issue creation and assignment.  If there's something that needs to be fixed, let's get it in an issue so we don't overlap.
1. Search to see if an issue exists
2. If it doesn't, create and assign yourself

It's tempting on smaller, quick issues like this to skip that but hopefully this prevents overlap.

**Comment by joelclimbsthings** [comment]
> Undoing dismiss and snooze is still broken in this branch.

@joshuatf  Just quickly tested dismiss and undo dismiss via the Homescreen task list, and all seems to work correctly with this branch? Just curious what you're referring to, although I haven't tested the snooze functionality on the front-end. 

Glad to defer to your PR ([7728](https://github.com/woocommerce/woocommerce-admin/pull/7728)) if you like. It looks like you may want to copy in the fix from this PR for `get_task()` in `TaskLists.php` when a `$task_list_id` is provided. Currently referencing a non-existent function. I'll jump over there to review as well.

**Comment by joelclimbsthings** [comment]
> I think the answer to this is issue creation and assignment. If there's something that needs to be fixed, let's get it in an issue so we don't overlap.
Search to see if an issue exists
If it doesn't, create and assign yourself

I would agree with this as well @joshuatf. Although I'd try to come up with something more concrete than "Search to see if an issue exists." Perhaps I'm just weak on my git search-foo, but it seems that it would still be easy to miss, and a bit tedious. For instances like this when we're working within an epic, adding the issue to the epic would make it easier to spot.

**Comment by joshuatf** [comment]
>@joshuatf Just quickly tested dismiss and undo dismiss via the Homescreen task list, and all seems to work correctly with this branch? Just curious what you're referring to, although I haven't tested the snooze functionality on the front-end.

I don't see them working at all on client-side.  The dismissal/snooze requests create a race condition and can leave the task invisible even after undoing the action.  In a nutshell:

1. Task gets snoozed
2. `isSnoozed` is optimistically updated to `true` in the data store
3. The task gets unsnoozed
4. `isSnoozed` is optimistically updated to `false` in the data store
2. Task is replaced with server-side task from success response of original snooze request and `is_visible` is `false`.
3. Task is no longer visible since `is_visible` was set to false

Part of this boils down to keeping visibility logic client-side.

>Perhaps I'm just weak on my git search-foo, but it seems that it would still be easy to miss, and a bit tedious. 

I searched for "task list" - generic enough and if it's a new issue it will generally be in the top 10 results.

https://github.com/woocommerce/woocommerce-admin/issues?q=is%3Aissue+is%3Aopen+task+list

>For instances like this when we're working within an epic, adding the issue to the epic would make it easier to spot.

Good idea.

**Comment by joelclimbsthings** [comment]
> I don't see them working at all on client-side. The dismissal/snooze requests create a race condition and can leave the task invisible even after undoing the action.

@joshuatf Your description makes sense to me, and I believe we should handle this, although it _does_ seem like a bit of an edge case? Your statement of "don't see them working at all on client-side" leads me to think I'm still not quite clear on this. Wouldn't it would only be consequential if the `undo_snooze` request returned before the initial `snooze` request? I wasn't able to reproduce this on this branch even if I clicked the "undo" button as quickly as possible.

That said, I do think we should cover this since it would be more common on mobile devices and flaky connections and all that.

image

**Comment by joshuatf** [comment]
>@joshuatf Your description makes sense to me, and I believe we should handle this, although it does seem like a bit of an edge case? 

I don't believe this is an edge case.  It occurs every time I test, including this branch.  Have you tested snoozing and unsnoozing on this branch?

I had to fix a few items to be able to see the snooze at it was initially missing.  After snoozing, I was not unable to unsnooze and the task was hidden until next page load due to the steps mentioned above.

Screen Shot 2021-09-29 at 7 56 24 AM

I think these issues are resolved in https://github.com/woocommerce/woocommerce-admin/pull/7728 so in the interest of time, I'd be in favor of going forward with that PR so we can hit our deadlines this week.

**Comment by joelclimbsthings** [comment]
> I don't believe this is an edge case. It occurs every time I test, including this branch. Have you tested snoozing and unsnoozing on this branch?

@joshuatf I had been testing dismissal since that was subject to the scenario you laid out above, without issue. You're right that snoozing wasn't working.

> I had to fix a few items to be able to see the snooze at it was initially missing. After snoozing, I was not unable to unsnooze and the task was hidden until next page load due to the steps mentioned above.

I saw the same behavior until I fixed a couple previous typos regarding the snoozing that you had fixed in your branch already. I don't think it really had anything to do with the scenario you laid out above. I would expect that that scenario would still be a potential issue on both of our branches, although an edge case as I remarked on earlier. It would only occur if the requests returned out of order. 

> It never hurts to also ask in Slack if there is any doubt.

@mattsherman True! That said, if we can find an async solution I think that would be a bit more reliable. 🤞🏻  

All that said, closing this issue since I just approved https://github.com/woocommerce/woocommerce-admin/pull/7728/ and it's ready to merge.

**Comment by mattsherman** [comment]
> @mattsherman True! That said, if we can find an async solution I think that would be a bit more reliable. 🤞🏻

Agreed. I think this would be a great thing for the team to discuss together. I'll set something up.
